include/onnxruntime/interface/graph/graph.h

```diff
@@ -0,0 +1,190 @@
+  /// </summary>
+  /// <param name="op_type">Op type</param>
+  /// <param name="domain">Domain. Empty string and "onnx.ai" are treated as equal.</param>
+  /// <returns></returns>
+  virtual bool IsOp(std::string_view op_type, std::string_view domain = "") const {
+    if (OpType() != op_type) {
+      return false;
+    }
+    std::string_view node_domain = Domain();
+    return node_domain == domain ||
+           ((domain == "" || domain == "ai.onnx") && (node_domain == "" || node_domain == "ai.onnx"));
+  }
+
+  /// <summary>
+  /// Convenience method. Returns value of int attribute with name, or given default if unset.
+  /// </summary>
+  /// <param name="name">Attribute name</param>
+  /// <param name="default_value">Default value</param>
+  /// <returns>Attribute value or default value</returns>
+  virtual int64_t GetAttributeIntDefault(std::string_view name, int64_t default_value) const {
+    return GetAttributeInt(name).value_or(default_value);
+  }
+
+  /// <summary>
+  /// Returns the schema since version for the op_type of this node. Value of -1 means it is not set.
+  /// </summary>
+  /// <returns>since version or default value -1</returns>
+  virtual int SinceVersion() const = 0;
+
+  // TODO: Shall we add this API to support subgraph access? looks it contradicts with the comment of GraphRef: No ability to access subgraphs is provided
+  virtual std::vector<std::unique_ptr<GraphViewRef>> GetSubgraphs() const = 0;
+
+  virtual ~NodeViewRef(){};
+};
+
+class GraphViewRef {
+ public:
+  /// <param name="domain">Domain name to find in model opset_import</param>
+  /// <returns>Opset of domain declared in model, or nullopt if domain is not present</returns>
+  virtual std::optional<int64_t> Opset(std::string_view domain) const = 0;
+
+  /// <returns>Topologically-sorted list of nodes in the graph</returns>
+  virtual std::vector<std::unique_ptr<NodeViewRef>> NodeViews() const = 0;
+
+  /// <summary>
+  /// Checks whether the value name refers to a constant initializer and if so, returns a Tensor corresponding to it.
+  /// Constants from parent graphs may be included.
+  /// </summary>
+  /// <param name="name">Value name. Must be nonempty.</param>
+  /// <returns>Tensor corresponding to the constant initializer or nullptr</returns>
+  virtual std::unique_ptr<TensorRef> GetConstant(std::string_view name) const = 0;
+
+  /// <summary>
+  /// Returns a ValueInfo instance for querying info about the value with the given name. Behavior is undefined if
+  /// the name does not refer to a value in the graph.
+  /// <param name="name">Value name. Must be nonempty.</param>
+  /// <returns>A ValueInfo instance corresponding to the value with the given name</returns>
+  virtual std::unique_ptr<ValueInfoViewRef> GetValueInfoView(std::string_view name) const = 0;
+
+  virtual std::unique_ptr<NodeViewRef> GetNodeViewProducingOutput(std::string_view name) const = 0;
```

Comment:
>GetNodeViewProducingOutput [](http://example.com/codeflow?start=39&length=26)

what is this?

Reply:
It is an API to get the node producing the specific NodeArg